File: WebLeave/SPA/src/app/_core/services/common/common.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { environment } from '@env/environment';
import { Area } from '@models/common/area';
import { Building } from '@models/common/building';
import { CommentArchive } from '@models/common/comment-archive';
import { Company } from '@models/common/company';
import { Department } from '@models/common/department';
import { BrowserInfo } from '@models/common/browser-info';

@Injectable({
  providedIn: 'root'
})
export class CommonService {

  apiUrl = environment.apiUrl;
  constructor(private http: HttpClient) { }
  getCompany() {
    return this.http.get<Company[]>(`${this.apiUrl}common/GetCompanys`);
  }
  getAreas() {
    return this.http.get<Area[]>(`${this.apiUrl}common/GetAreas`);
  }
  getBuildings() {
    return this.http.get<Building[]>(`${this.apiUrl}common/GetBuildings`);
  }
  getDepartments() {
    return this.http.get<Department[]>(`${this.apiUrl}common/GetDepartments`);
  }
  getCommentArchives() {
    return this.http.get<CommentArchive[]>(`${this.apiUrl}common/GetCommentArchives`);
  }
  getBrowserInfo(ipLocal: string, username: string) {
    return this.http.get<BrowserInfo>(`${this.apiUrl}common/GetBrowserInfo`, { params: { ipLocal, username } });
  }

  getServerTime() {
    return this.http.get<Date>(`${this.apiUrl}common/GetServerTime`);
  }
}

File: WebLeave/API/_Services/Services/Common/CommonService.cs
using API._Repositories;
using API._Services.Interfaces.Common;
using API.Dtos.Common;
using API.Models;
using Microsoft.EntityFrameworkCore;
namespace API._Services.Services.Common
{
    public class CommonService : ICommonService
    {
        private readonly IRepositoryAccessor _repoAccessor;
        public CommonService(IRepositoryAccessor repoAccessor)
        {
            _repoAccessor = repoAccessor;
        }

        public async Task<List<Area>> GetAreas()
        {
            return await _repoAccessor.Area.FindAll().ToListAsync();
        }

        public async Task<List<Building>> GetBuildings()
        {
            return await _repoAccessor.Building.FindAll().ToListAsync();
        }

        public async Task<List<CommentArchive>> GetCommentArchives()
        {
            return await _repoAccessor.CommentArchive.FindAll().OrderBy(x => x.Value).ToListAsync();
        }

        public async Task<List<Company>> GetCompanys()
        {
            return await _repoAccessor.Company.FindAll().ToListAsync();
        }

        public async Task<List<Department>> GetDepartments()
        {
            return await _repoAccessor.Department.FindAll().ToListAsync();
        }

        public async Task<BrowserInfoDto> GetLoginDetectInfo(string username)
        {
            BrowserInfoDto result = new()
            {
                Factory = SettingsConfigUtility.GetCurrentSettings("AppSettings:Factory")
            };
            if (!string.IsNullOrEmpty(username?.Trim()))
                result.LoginDetect = await _repoAccessor.LoginDetect.FirstOrDefaultAsync(x => x.UserName == username.Trim());

            return result;
        }

        public DateTime GetServerTime()
        {
            // Lấy múi giờ của máy chủ  
            var serverTimeZone = TimeZoneInfo.Local;
            var utcNow = DateTime.UtcNow;
            // Chuyển đổi sang thời gian máy chủ 
            return TimeZoneInfo.ConvertTimeFromUtc(utcNow, serverTimeZone);
        }
    }
}